liblwgeom/lwgeom_median.c

```diff
@@ -19,156 +19,247 @@
+	assert(denom >= 0);
 
-	/* If any of the intermediate points in the calculation is found in the
-	 * set of input points, the standard Weiszfeld method gets stuck with a
-	 * divide-by-zero.
-	 *
-	 * To get ourselves out of the hole, we follow an alternate procedure to
-	 * get the next iteration, as described in:
-	 *
-	 * Vardi, Y. and Zhang, C. (2011) "A modified Weiszfeld algorithm for the
-	 * Fermat-Weber location problem."  Math. Program., Ser. A 90: 559-566.
-	 * DOI 10.1007/s101070100222
-	 *
-	 * Available online at the time of this writing at
-	 * http://www.stat.rutgers.edu/home/cunhui/papers/43.pdf
-	 */
-	if (hit)
+	/* denom is zero in case of multipoint of single point when we've converged perfectly */
+	if (denom > DBL_EPSILON)
 	{
-		double dx = 0;
-		double dy = 0;
-		double dz = 0;
-		double r_inv;
-		POINT3D alt;
-		for (i = 0; i < npoints; i++)
+		next.x /= denom;
+		next.y /= denom;
+		next.z /= denom;
+
+		/* If any of the intermediate points in the calculation is found in the
+	 	* set of input points, the standard Weiszfeld method gets stuck with a
+	 	* divide-by-zero.
+	 	*
+	 	* To get ourselves out of the hole, we follow an alternate procedure to
+	 	* get the next iteration, as described in:
+	 	*
+	 	* Vardi, Y. and Zhang, C. (2011) "A modified Weiszfeld algorithm for the
+	 	* Fermat-Weber location problem."  Math. Program., Ser. A 90: 559-566.
+	 	* DOI 10.1007/s101070100222
+	 	*
+	 	* Available online at the time of this writing at
+	 	* http://www.stat.rutgers.edu/home/cunhui/papers/43.pdf
+	 	*/
+		if (hit)
 		{
-			if (distances[i] > 0)
+			double dx = 0;
+			double dy = 0;
+			double dz = 0;
+			double r_inv;
+			for (i = 0; i < npoints; i++)
 			{
-				dx += (points[i].x - curr->x) / distances[i];
-				dy += (points[i].y - curr->y) / distances[i];
-				dz += (points[i].y - curr->z) / distances[i];
+				if (distances[i] > DBL_EPSILON)
+				{
+					dx += (points[i].x - curr->x) / distances[i];
+					dy += (points[i].y - curr->y) / distances[i];
+					dz += (points[i].y - curr->z) / distances[i];
+				}
 			}
-		}
-
-		r_inv = 1.0 / sqrt ( dx*dx + dy*dy + dz*dz );
 
-		alt.x = FP_MAX(0, 1.0 - r_inv)*next.x + FP_MIN(1.0, r_inv)*curr->x;
-		alt.y = FP_MAX(0, 1.0 - r_inv)*next.y + FP_MIN(1.0, r_inv)*curr->y;
-		alt.z = FP_MAX(0, 1.0 - r_inv)*next.z + FP_MIN(1.0, r_inv)*curr->z;
+			r_inv = 1.0 / sqrt ( dx*dx + dy*dy + dz*dz );
 
-		next = alt;
-	}
+			next.x = FP_MAX(0, 1.0 - r_inv)*next.x + FP_MIN(1.0, r_inv)*curr->x;
+			next.y = FP_MAX(0, 1.0 - r_inv)*next.y + FP_MIN(1.0, r_inv)*curr->y;
+			next.z = FP_MAX(0, 1.0 - r_inv)*next.z + FP_MIN(1.0, r_inv)*curr->z;
+		}
 
-	delta = distance3d_pt_pt(curr, &next);
+		delta = distance3d_pt_pt((POINT3D*)curr, (POINT3D*)&next);
 
-	curr->x = next.x;
-	curr->y = next.y;
-	curr->z = next.z;
+		curr->x = next.x;
+		curr->y = next.y;
+		curr->z = next.z;
+	}
 
 	return delta;
 }
 
-static POINT3D
-init_guess(const POINT3D* points, uint32_t npoints)
+static POINT4D
```

Comment:
Why is this returning a `POINT4D` ? What is the significance of the `m` ordinate on the returned value?

Reply:
changed